File: samples/SampleCounter/js/src/sample-counter/counter-presenter.ts
import { Counter } from "./counter";
import type { ICounterView } from "./i-counter-view";

export class CounterPresenter {
  private readonly _view: ICounterView;

  private _counter: Counter = Counter.zero;

  constructor(view: ICounterView) {
    this._view = view;
    this.initialize();
  }

  private initialize(): void {
    this._view.displayCounter(this._counter);
  }

  increment(): void {
    this._counter = this._counter.increment();
    this.displayCounter();
  }

  decrement(): void {
    this._counter = this._counter.decrement();
    this.displayCounter();
  }

  private displayCounter(): void {
    this._view.displayCounter(this._counter);
  }
}

File: targets/flutter/sample_counter/lib/sample_counter/counter_presenter.dart
import 'counter.dart';
import 'i_counter_view.dart';

final class CounterPresenter {
  CounterPresenter(ICounterView view) {
    this._view = view;
    this.initialize();
  }

  late final ICounterView _view;

  Counter _counter = Counter.zero;

  void initialize() => this._view.displayCounter(this._counter);

  void increment() {
    this._counter = this._counter.increment();
    this.displayCounter();
  }

  void decrement() {
    this._counter = this._counter.decrement();
    this.displayCounter();
  }

  void displayCounter() => this._view.displayCounter(this._counter);
}
